sklearn/gaussian_process/kernels.py

```diff
@@ -1617,6 +1617,132 @@ def __repr__(self):
             self.__class__.__name__, self.length_scale, self.periodicity)
 
 
+class LinearKernel(Kernel):
+    """Linear kernel.
+
+    The Linear kernel is non-stationary and can be obtained from Bayesian
+    linear regression by shifting the origin by c. The parameters of the Linear
+    kernel are about specifying the origin. The kernel is given by:
+
+    k(x_i, x_j) = ((x_i - c) \cdot (x_j - c))
+
+    The Linear kernel can be combined with other kernels, more commonly with
+    periodic kernels.
+
+    Parameters
+    ----------
+    c : float or array with shape (n_features,), default: 0.0
+        The offset for the origin. If a float, the same offset is used
+        throughout all dimensions. If an array, each dimension of c defines the
+        offset of the respective feature dimension.
+
+    c_bounds : pair of floats >= 0, default: (1e-5, 1e5)
+        The lower and upper bound on c
+    """
+    def __init__(self, c=0.0, c_bounds=(1e-5, 1e5)):
+        self.c = c
+        self.c_bounds = c_bounds
+
+    @property
+    def non_uniform_offset(self):
+        return np.iterable(self.c) and len(self.c) > 1
+
+    @property
+    def hyperparameter_c(self):
+        if self.non_uniform_offset:
+            return Hyperparameter("c", "numeric", self.c_bounds, len(self.c))
+        else:
+            return Hyperparameter("c", "numeric", self.c_bounds)
+
+    def __call__(self, X, Y=None, eval_gradient=False):
+        """Return the kernel k(X, Y) and optionally its gradient.
+
+        Parameters
+        ----------
+        X : array, shape (n_samples_X, n_features)
+            Left argument of the returned kernel k(X, Y)
+
+        Y : array, shape (n_samples_Y, n_features), (optional, default=None)
+            Right argument of the returned kernel k(X, Y). If None, k(X, X)
+            if evaluated instead.
+
+        eval_gradient : bool (optional, default=False)
+            Determines whether the gradient with respect to the kernel
+            hyperparameter is determined. Only supported when Y is None.
+
+        Returns
+        -------
+        K : array, shape (n_samples_X, n_samples_Y)
+            Kernel k(X, Y)
+
+        K_gradient : array (opt.), shape (n_samples_X, n_samples_X, n_dims)
+            The gradient of the kernel k(X, X) with respect to the
+            hyperparameter of the kernel. Only returned when eval_gradient
+            is True.
+        """
+        X = np.atleast_2d(X)
+        self.c = _check_length_scale(X, self.c)
+        if Y is None:
+            K = np.inner(X - self.c, X - self.c)
```

Comment:
We might as well avoid doing this subtraction repeatedly. It is done below again in c_gradient computation. Perhaps we should even avoid subtraction if c==0